core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,324 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.StorageTierAssoc;
+import alluxio.WorkerStorageTierAssoc;
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.InvalidWorkerStateException;
+import alluxio.exception.PreconditionMessage;
+import alluxio.exception.WorkerOutOfSpaceException;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.network.NetworkAddressUtils;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.io.LocalFileBlockWriter;
+import alluxio.worker.block.meta.UnderFileSystemBlockMeta;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class implements a {@link BlockReader} to read a block directly from UFS, and optionally
+ * cache the block to the Alluxio worker if the whole block it is read.
+ */
+@NotThreadSafe
+public final class UnderFileSystemBlockReader implements BlockReader {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemBlockReader.class);
+
+  /** An object storing the mapping of tier aliases to ordinals. */
+  private final StorageTierAssoc mStorageTierAssoc = new WorkerStorageTierAssoc();
+
+  /** The initial size of the block allocated in Alluxio storage when the block is cached. */
+  private final long mInitialBlockSize;
+  /** The block metadata for the UFS block. */
+  private final UnderFileSystemBlockMeta mBlockMeta;
+  /** If set, do not cache the block. */
+  private final boolean mNoCache;
+  /** The Alluxio block store. It is used to interact with Alluxio. */
+  private final BlockStore mAlluxioBlockStore;
+
+  /** The input stream to read from UFS. */
+  private InputStream mUnderFileSystemInputStream;
+  /** The block writer to write the block to Alluxio. */
+  private LocalFileBlockWriter mBlockWriter;
+  /** If set, the reader is closed and should not be used afterwards. */
+  private boolean mClosed;
+
+  /**
+   * The position of mUnderFileSystemInputStream (if not null) is blockStart + mInStreamPos.
+   * When mUnderFileSystemInputStream is not set, this is set to -1 (an invalid state) to indicate
+   * that the input stream is not initialized.
+   */
```

Comment:
How about checking the inputstream before accessing or returning `mInStreamPos`?